### PR TITLE
Simplify flattening

### LINKED ISSUE #75: Stricter logic for `cdm_flatten_to_tbl()`
The function itself should not think too much by itself, it should apply the same logic for all join-methods. Furthermore, only tables that are (direct) parents will be allowed for this function (no grandparents etc.). If the user passes tables in the ellipsis, the joins will happen in this order.

1. `left_join()`, `inner_join()`, `full_join()`: if no tables given in the ellipsis, the result will depend on the order of the tables in the (relevant part of the) `dm`, but it will only affect the column order.
2. `semi_join()`: the result will be exactly those rows, that are referenced by ALL parent tables.
3. `anti_join()`: the result will be those rows, that are NOT referenced by ANY parent table (i.e. referenced by NO parent table). It might be more interesting for the user to have a function telling them, which rows are not referenced by at least one parent table, but that needs to be another function.
4. `nest_join()`: the created columns should have the names of the RHS tables of the join. Again, if ellipsis empty, the order of those tables in the `dm` plays a role for the resulting order of the columns.
5. `right_join()`: if the ellipsis is empty, the number of rows might depend on the table that comes last in the `dm` table list (if no referential integrity is given). In this case, a warning is issued.

There will be a further function `cdm_flatten_deep_to_tbl()`, which works only for `left_join()`, `inner_join()` and `full_join()`, which also allows for deeper table hierarchies (grandparents etc.)

### PR DESCRIPTION
**Pull request by krlmlr**
@TSchiefer: Can you please pick up from here when next working on it?

Closes #75.

### CODE REVIEW

**Comment by TSchiefer** [comment]
@krlmlr 
State-of-the-Art:

1. `join = nest_join` leads to an error, cause it is a useless action (and might waste time)
1. `right_join` and `full_join` on a filtered `dm` lead to an error (tests supported theory of #75)
1. `cdm_flatten_to_tbl()`:
    - documentation is updated
    - tests
    - fails when "grandparent" tables are involved
1. `cdm_squash_to_tbl()`:
    - works for "grandparent" tables when `join %in% c(left_join, full_join, inner_join)`
    - tests

Could you please review? (requesting doesn't work in this case, maybe cause you started the branch?)

**Comment by krlmlr** [comment]
Can you please resolve conflicts and merge?

**Comment by krlmlr** [comment]
Thanks!

File: R/flatten.R [hunk 1]
```diff
@@ -154,33 +148,39 @@ cdm_flatten_to_tbl_impl <- function(dm, start, ..., join, join_name) {
...
-  # Need to use `dm` and not `clean_dm` here, cause of possible filter conditions.
-  start_tbl <- tbl(dm, start) %>% rename(!!!renames)
-
   # Drop first table in the list of join partners. (We have at least one table, `start`.)
   # (Working with `reduce2()` here and the `.init`-parameter is the first table)
   # in the case of only one table in the `dm` (table "start"), all code below is a no-op
   order_df <- order_df[-1, ]
+  # the order given in the ellipsis determines the join-list
+  if (!auto_detect) {
+    pt_names <- order_df$name
```
[the remark this answers is not in the document]
Comment (TSchiefer) [reply R/flatten.R]: argh, just imagine: I did mess up the order of the arguments in `match()`... (I guess the tests passed purely by chance) good you asked for a more transparent implementation.

[the remark this answers is not in the document]
Comment (krlmlr) [reply R/flatten.R]: No worries. Is there a way to split `cdm_flatten_to_tbl_impl()` into separate functions which could be tested individually?

**Pull request merged**